src/storage/src/source/mysql/snapshot.rs

```diff
@@ -407,14 +407,30 @@ pub(crate) fn render<G: Scope<Timestamp = GtidPartition>>(
                 let mut final_row = Row::default();
 
                 let mut snapshot_staged = 0;
-                for (table, (output_index, table_desc)) in reader_snapshot_table_info {
+                'outer: for (table, (output_index, table_desc)) in reader_snapshot_table_info {
                     let query = format!("SELECT * FROM {}", table.to_ast_string());
                     trace!(%id, "timely-{worker_id} reading snapshot from \
                                  table '{table}':\n{table_desc:?}");
                     let mut results = tx.exec_stream(query, ()).await?;
                     while let Some(row) = results.try_next().await? {
                         let row: MySqlRow = row;
-                        let packed_row = pack_mysql_row(&mut final_row, row, &table_desc)?;
+                        let packed_row = match pack_mysql_row(&mut final_row, row, &table_desc) {
+                            Ok(row) => row,
+                            Err(err @ MySqlError::ValueDecodeError { .. }) => {
+                                // If we hit an error decoding there's nothing more we can do
+                                let err = DefiniteError::ValueDecodeError(err.to_string());
+                                raw_handle
+                                    .give(
+                                        &data_cap_set[0],
+                                        ((output_index, Err(err)), GtidPartition::minimum(), 1),
+                                    )
+                                    .await;
+                                trace!(%id, "timely-{worker_id} stopping snapshot of table \
+                                             {table} due to decode error");
```

Comment:
The same treatment should happen here, the snapshot should keep going and all definite errors should be emitted into the dataflow in the exact same way as normal data. If the problematic rows are later removed from the table then the remaining non-errored rows will be revealed and those should represent the correct state of the table.

Reply:
>  If the problematic rows are later removed from the table then the remaining non-errored rows will be revealed and those should represent the correct state of the table.

Ah okay -- I didn't realize that was possible! 

How would things work if the contents of the error are the same for multiple bad rows in the table? If any of them are later deleted, this would mean we might retract the error value of the wrong row? Is this why it's important to have a text-version of the failing row included in the error, to make it unique?

Reply:
> If any of them are later deleted, this would mean we might retract the error value of the wrong row?

If there are two invalid rows then we will emit an `(error, gtid, +1)` update **for each of them**. Which means that the total diff for that error will be 2. These two copies are fungible (because they are identical) and as such it doesn't make sense to talk about "the error of the wrong row". There are just two copies of the same error and if one invalid row is removed then we will just emit `(error, gtid, -1)` and reduce the total accumulated diff to `1`. The source will sill remain in an errored state until the second copy is removed.

>  Is this why it's important to have a text-version of the failing row included in the error, to make it unique?

No this is just to make it easier for users to know which row in their original data is the problematic one

Reply:
Ok cool that makes sense!

Reply:
Unfortunately it looks like we can't actually retract DefiniteErrors -- doing so causes a panic `unexpected retraction of definite error`.

Looks like this was introduced by @sploiselle to introduce the ceased status here:
https://github.com/MaterializeInc/materialize/commit/60290f4f5aafddb057111d5aa60c30c192732fb9

I'm unsure of the implications on changing that, since it does seem that it'd be a valid scenario to allow deletion of a problematic row to get a source back into a working state.

Reply:
I think this PR is also related: https://github.com/MaterializeInc/materialize/pull/24531/

In that PR we emit positive diffs for all DefiniteErrors, even if they were originally produced as a retraction. I think that behavior makes sense for breaking schema changes which should be _definite_, but in this case of a single row failing to decode I'm not sure that it does. If I leave things as-is (a `TransientError`), it'll cause the operator to be restarted each time it encounters the failing row and stall the entire source.

Reply:
>  but in this case of a single row failing to decode I'm not sure that it does.

I agree, we need to adjust the status handling to not have preferences on diffs of errors. I wonder if the misunderstanding comes from using the word "definite" to describe the error which can be confused with permanent? In any case, we definitely want definite errors to be retractable.

I think the correct way to do this is what I had described in bullet 3 here https://github.com/MaterializeInc/materialize/pull/23945#discussion_r1431460039 . If the "ceased" state needs to communicate finality then frontiers are the way to be sure about it.